core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -353,6 +353,17 @@ class ZooKeeperClient(connectString: String,
     info("Closed.")
   }
 
+  /**
+   * Wait up to the provided timeoutMs this ZooKeeperClient to close.
+   * Supplying a timeoutMs of 0 means to wait forever.
+   */
+  def closeAndWait(timeoutMs: Int): Boolean = {
```

Comment:
Is there any benefit to wait for the ZK threads to finish? ZooKeeper.close() calls cnxn.close(), which calls disconnect(), which waits for sendThread to finish. So, the only thread not closed yet in ZooKeeper is eventThread. However, once ZooKeeper.close() is called, the eventThread is useless since any new event will fail.

Reply:
The motivation for this PR is to fail specific tests which seem to leave a ZK event thread running (which has caused test failures in the past due to the `verifyNoUnexpectedThreads` check here https://github.com/apache/kafka/blob/91bbaac9bd42d79e03c87c8903b53052fabf8bc6/core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala#L123).

I'm not sure on the motivation for `verifyNoUnexpectedThreads` beyond ensuring resources are cleaned up between tests. To you point, a single orphaned `eventThread` seems harmless, but it may cause problems if tests are repeatedly creating these threads without cleaning them up.

Reply:
@junrao @gardnervickers The goal of that check is to identify clients still running (consumer, producer, admin client, zkclient, etc.) because they cause other tests to fail in unpredictable ways (metadata requests may cause topics to be created, zk requests can cause the security context to be loaded before it's set, etc.).

We could change the check to check for the send thread instead, but in our own clients, we do tend to wait for our threads to complete before we consider `close` done.

Reply:
Thanks for the explanation. In that case, could we just change ZooKeeperClient.close() to wait for the ZK SendThread to finish? In our usage, the calls from the SendThread just append to an unbounded queue. So, SendThread should finish very quickly after close(). We also call zooKeeper.close() in reinitialize(). We probably want to make that call wait for the SendThread too.

Reply:
Sounds good.